File: Recursion/4.2-Count-Using-Recursion/c/code.c
#include <stdio.h>

int length(int arr[], int idx)
{
    if (arr[idx] == '\0')
    {
        return 0;
    }

    return 1 + length(arr, idx + 1);
}

int main()
{
    int a[] = {1, 2, 3, 4, 5};

    int result = length(a, 0);

    printf("Length:%d\n", result);
    return 0;
}

File: Recursion/4.4-Binary-Search-Using-Recursion/go/code.go
package main

import "fmt"

func search(arr []int, target int) int {

	length := len(arr)

	if length == 0 {
		return -1
	}

	var mid int = (length - 1) / 2
	var guess int = arr[mid]

	if guess == target {
		return mid
	}
	if target < guess {
		position := search(arr[:mid], target)
		if position == -1 {
			return -1
		}
		return position
	} else {
		position := search(arr[mid+1:], target)
		if position == -1 {
			return -1
		}
		return mid + 1 + position
	}

}

func main() {
	arr := make([]int, 10)

	for i := 0; i < 10; i++ {
		arr[i] = i + 1
	}

	for i := 10; i > 0; i-- {
		result := search(arr, i)

		fmt.Printf("%d found at: %d\n", i, result)
	}
}
